pkg/server/singleprocess/service_pipeline.go

```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
 			}
 		}
 
-		run.Jobs = append(run.Jobs, &pb.Ref_Job{Id: job.Id})
+		pipelineRun.Jobs = append(pipelineRun.Jobs, &pb.Ref_Job{Id: job.Id})
 		stepJobs = append(stepJobs, &pb.QueueJobRequest{
 			Job: job,
 		})
 	}
 
-	// Get the graph for the steps so we can get the root. We enforce a
-	// single root so the root is always the first step.
-	stepGraph, err := serverptypes.PipelineGraph(pipeline)
-	if err != nil {
-		return nil, err
+	return stepJobs, pipelineRun, stepIds, nil
+}
+
+// pipelineGraphFull takes a pipeline, and optionally accepts an existing Graph
+// and parent step, and attempts to build a full graph for a given Pipeline including
+// any nested pipeline steps. It keeps track of visisted steps so that we don't
+// get stuck in a loop building the graph. This means step names must be unique
+// across pipelines.
+func (s *Service) pipelineGraphFull(
+	ctx context.Context,
+	log hclog.Logger,
+	g *graph.Graph,
+	parentStep string,
+	visistedNodes map[string]string,
+	nodeIdMap map[string]*pb.Ref_PipelineStep,
+	pipeline *pb.Pipeline,
+) (*graph.Graph, map[string]*pb.Ref_PipelineStep, error) {
+	var stepGraph *graph.Graph
+	if g != nil {
+		// We're handling an embedded pipeline graph
+		stepGraph = g
+	} else if stepGraph == nil {
+		stepGraph = &graph.Graph{}
 	}
 
-	// Get the ordered jobs.
-	var jobIds []string
-	jobMap := map[string]*pb.Ref_PipelineStep{}
-	for _, v := range stepGraph.KahnSort() {
-		jobId := stepIds[v.(string)]
-		jobIds = append(jobIds, jobId)
-		jobMap[jobId] = &pb.Ref_PipelineStep{
-			Pipeline:    pipeline.Id,
-			Step:        v.(string),
-			RunSequence: run.Sequence,
+	// Note that pipeline.Steps is not an ordered list of steps. It's a map of key val
+	// steps so the order will not match the order steps are defined in a waypoint.hcl.
+	for _, step := range pipeline.Steps {
+		if len(visistedNodes) != 0 {
+			if pipeName, ok := visistedNodes[step.Name]; ok && pipeName == pipeline.Name {
```

Comment:
```suggestion
			if pipeName, ok := visitedNodes[step.Name]; ok && pipeName == pipeline.Name {
```
lol Brian. After seeing this word so many times even I'm like, "wait, how do you spell visited? Is visisted the actual word?"